docs/CppInteroperability/ForwardVision.md

```diff
@@ -0,0 +1,290 @@
+
+One of the most common uses of this "API patterns" concept concerns the import of types. Swift types fall into two categories: value types and reference types. Copying a value of a reference type produces a new reference to the same underlying object, similar to an intrusive `std::shared_ptr` in C++ or a class type in Java. Copying a value of a value type recursively copies the components of the type, similar to the behavior of a struct in C or the default behavior of a class type in C++. Furthermore, both kinds of types must always be copyable, although there are plans in the works to allow types to restrict this.
+
+Types in C++ do not always fit cleanly into this model, and they cannot always be automatically mapped to it even when they do. As discussed above, many C++ class types are idiomatically used as reference types; they are always passed around as a reference, either using a raw pointer (`*`) or reference (`&`) type or using a smart pointer type such as `std::shared_ptr` or `std::unique_ptr`. C++ class types can also customize or even remove their value operations. Some types that do this still have "value semantics": they function as self-contained values with referential transparency. Others maintain a more hybrid semantics, or have external dependencies, or make themselves uncopyable or even unmovable, or are even just using types as a language mechanism for getting a scoped destructor.
+
+The "API patterns" model picks off various C++ APIs that Swift can handle in a specific, idomatic way such as self contained value types, reference types, singletons, iterators, generalized accessors, and so on.
+
+### Reference types
+
+Like Swift types, Objective-C types fall into two categories, which makes importing them easy: structs are imported as value types and Objective-C classes are imported as Swift class types. The same is not true for C++. As discussed above, there is no clear idiom for defining reference types in C++. All types in C++ are declared in the same way, but some types have reference semantics and others don’t. To be able to express reference types in Swift natively, users must annotate their reference types as such, which will tell the compiler to import them as Swift classes (which have the same semantics).
+
+When importing a reference type, there are three main sub-categories that reference type can fall into: 1) **immortal** reference types are “intentionally leaked” and live for the duration of the program 2) **manually managed** reference types are constructed and destroyed manually by the programmer, these are unsafe reference types and should maybe be renamed accordingly 3) **automatically managed** reference types provide retain and release operations which are used to safely destroy a value of the type after its last use. [(Examples below.)](## Examples and Definitions)
```

Comment:
```suggestion
There are several common patterns for reference types in C++. Swift's C++ interoperability should support at least these three:
- **Immortal** reference types are not designed to be managed individually by the program. Objects of these types are allocated and then intentionally "leaked" without tracking their uses.

  Sometimes these objects are not truly immortal: for example, they may be arena-allocated, with an expectation that they will only be referenced from other objects within the arena. Nonetheless, they aren't expected to be individually managed.
- **Manually managed** reference types are constructed and destroyed manually by the programmer.  Usually they're uniquely referenced and there's some sort of explicit operation to destroy them, either the standard `delete` operator or a specific method.
- **Automatically managed** reference types are reference-counted with retain and release operations.
[Examples of each of these are given below.](## Examples and Definitions)
```

I have to say that I think these last two are poorly-chosen names.  I think you're talking about unique reference types and shared reference types, and there's an orthogonal dimension of whether or not the API expects the programmer to manage ownership explicitly.